File: 백준/Silver/1018. 체스판 다시 칠하기/체스판 다시 칠하기.cc
#include <iostream>

using namespace std;

int min(int a, int b) {
	if (a >= b) return b;
	return a;
}

int main() {
	char graph[51][51];
	int n, m;
	cin >> n >> m;

	for (int i = 0; i < n; i++) {
		for (int j = 0; j < m; j++) {
			cin >> graph[i][j];
		}
	}

	char target1[51][51];
	char target2[51][51];

	for (int i = 0; i < 8; i++) {
		for (int j = 0; j < 8; j++) {
			if ((i+j) % 2 == 0) {
				target1[i][j] = 'W';
				target2[i][j] = 'B';
			}
			else {
				target1[i][j] = 'B';
				target2[i][j] = 'W';
			}
		}
	}
	
	int ans = 50*50;
	int first, second;
	
	for (int i = 0; i <= n-8; i++) {
		for (int j = 0; j <= m-8; j++) {
			int cnt;
			first = second = 0;
			for (int y = i; y < i+8; y++) {
				for (int x = j; x < j+8; x++) {
					if (graph[y][x] != target1[y-i][x-j]) first++;
					if (graph[y][x] != target2[y-i][x-j]) second++;
				}
			}
			cnt = min(first, second);
			ans = min(cnt, ans);
		}
	}

	cout << ans;
}

File: 백준/Silver/1018. 체스판 다시 칠하기/체스판 다시 칠하기.java
import java.io.*;
import java.util.*;

public class Main {
    static int n, m;
    static char[][] graph;
    static String[] compareString;

    static int compare(int j, int i) {
        int result = Integer.MAX_VALUE;
        for (int k = 0; k < 2; k++) {
            int index = k;
            int cnt = 0;

            for (int y = i; y < i+8; y++) {
                for (int x = j; x < j+8; x++) {
                    if (graph[y][x] != compareString[index % 2].charAt(x-j)) cnt++;
                }
                index++;
            }

            result = Math.min(result, cnt);
        }

        return result;
    }

    public static void main(String[] args) throws IOException {
        // 입력 최적화
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        
        // 입력값 초기화
        StringTokenizer st = new StringTokenizer(br.readLine());
        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());
        graph = new char[n][m];

        for (int i = 0; i < n; i++) {
            String tmp = br.readLine();
            for (int j = 0; j < m; j++) graph[i][j] = tmp.charAt(j);
        }

        // 비교 문자열 초기화
        compareString = new String[2];
        compareString[0] = "WBWBWBWB";
        compareString[1] = "BWBWBWBW";

        // 로직
        int answer = Integer.MAX_VALUE;
        for (int i = 0; i <= n-8; i++) {
            for (int j = 0; j <= m-8; j++) {
                answer = Math.min(answer, compare(j, i));
            }
        }

        System.out.println(answer);
    }
}